File: packages/sync/src/sync-model.ts
export const MIRROR_ENGINE_PROFILE = "exact_document_plan_only_v1" as const;
export const MIRROR_PLANNER_POLICY = "three_way_exact_document_v1" as const;
export const MIRROR_PROJECTION_POLICY = "portable_mirror_projection_v1" as const;

export type SyncObjectKind = "record" | "resource" | "file";
export type SyncPlanReason = "initial" | "rebuild" | "local_change" | "remote_change" | "pending";

export interface SyncObjectRef {
  entity: SyncObjectKind;
  identity: string;
  path: string;
  revision: string;
  /** Exact bytes capability: document revision for text, content digest for files. */
  payload_revision: string;
  size?: number;
}

/** Absence is an explicit compare-and-swap state, never a missing check. */
export type ExpectedObjectState =
  | { state: "absent" }
  | { state: "exact"; object: SyncObjectRef };

export interface SyncCheckpoint {
  generation: number;
  cursor: number | null;
}

export interface SyncPlanBoundary {
  engine_profile: typeof MIRROR_ENGINE_PROFILE;
  protocol_profile: "exact_document_v1";
  planner_policy: typeof MIRROR_PLANNER_POLICY;
  projection_policy: typeof MIRROR_PROJECTION_POLICY;
  replica_id: string;
  scope_epoch: number;
  authority_cursor: number;
  checkpoint: SyncCheckpoint;
  selective_sync_fingerprint: string;
}

interface SyncActionBase {
  action_id: string;
  depends_on: string[];
  reason: SyncPlanReason;
}

export interface WriteLocalAction extends SyncActionBase {
  command: "write_local";
  target: SyncObjectRef;
  payload_revision: string;
  expected_local: ExpectedObjectState;
  expected_path_owner: ExpectedObjectState;
}

export interface DeleteLocalAction extends SyncActionBase {
  command: "delete_local";
  target: SyncObjectRef;
  expected_local: ExpectedObjectState;
  expected_path_owner: ExpectedObjectState;
}

export interface MoveLocalAction extends SyncActionBase {
  command: "move_local";
  source: SyncObjectRef;
  target_path: string;
  expected_source_owner: ExpectedObjectState;
  expected_target_owner: ExpectedObjectState;
}

export interface PutRemoteAction extends SyncActionBase {
  command: "put_remote";
  target: SyncObjectRef;
  payload_revision: string;
  expected_remote: ExpectedObjectState;
  expected_local: ExpectedObjectState;
  idempotency_key: string;
}

export interface DeleteRemoteAction extends SyncActionBase {
  command: "delete_remote";
  target: SyncObjectRef;
  expected_remote: ExpectedObjectState;
  expected_local: ExpectedObjectState;
  idempotency_key: string;
}

export interface MoveRemoteAction extends SyncActionBase {
  command: "move_remote";
  source: SyncObjectRef;
  target_path: string;
  expected_source_owner: ExpectedObjectState;
  expected_target_owner: ExpectedObjectState;
  expected_local: ExpectedObjectState;
  /** Opaque authority revisions produced by a prior command are receipt-bound. */
  revision_from_dependency?: string;
  idempotency_key: string;
}

export interface RecordConflictAction extends SyncActionBase {
  command: "record_conflict";
  identity: string;
  entity: "record" | "file";
  local: ExpectedObjectState;
  remote: ExpectedObjectState;
  conflict_kind:
    | "both_changed"
    | "delete_vs_change"
    | "path_occupied"
    | "rejected";
}

export interface AdvanceCheckpointAction extends SyncActionBase {
  command: "advance_checkpoint";
  expected: SyncCheckpoint;
  next: SyncCheckpoint;
}

export type SyncAction =
  | WriteLocalAction
  | DeleteLocalAction
  | MoveLocalAction
  | PutRemoteAction
  | DeleteRemoteAction
  | MoveRemoteAction
  | RecordConflictAction
  | AdvanceCheckpointAction;

export type SyncBatchPhase =
  | "prepared"
  | "applying"
  | "effects_complete"
  | "cancelled"
  | "blocked";

export interface SyncFailure {
  code: string;
  message: string;
  action_id?: string;
  precondition?: string;
}

export type SyncOutcomeStatus =
  | "applied"
  | "attention"
  | "cancelled"
  | "stale"
  | "blocked"
  | "failed";

export function isEffectAction(
  action: SyncAction
): action is Exclude<SyncAction, AdvanceCheckpointAction | RecordConflictAction> {
  return action.command !== "advance_checkpoint" && action.command !== "record_conflict";
}

File: crates/connect-mirror/src/sync_model.rs
use mdbase_connect_protocol::{SelectiveSyncPolicy, SyncReplicaMode};
use serde::{Deserialize, Serialize};

pub const ENGINE_PROFILE: &str = "exact_document_plan_only_v1";
pub const PROTOCOL_PROFILE: &str = "exact_document_v1";
pub const PLANNER_POLICY: &str = "three_way_exact_document_v1";
pub const PROJECTION_POLICY: &str = "portable_mirror_projection_v1";

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
#[serde(rename_all = "snake_case")]
pub enum SyncObjectKind {
    Record,
    Resource,
    File,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SyncObjectRef {
    pub entity: SyncObjectKind,
    pub identity: String,
    pub path: String,
    pub revision: String,
    pub payload_revision: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub size: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(tag = "state", rename_all = "snake_case")]
pub enum ExpectedObjectState {
    Absent,
    Exact { object: SyncObjectRef },
}

impl ExpectedObjectState {
    pub fn exact(&self) -> Option<&SyncObjectRef> {
        match self {
            Self::Absent => None,
            Self::Exact { object } => Some(object),
        }
    }
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum SyncPlanReason {
    Initial,
    Rebuild,
    LocalChange,
    RemoteChange,
    Pending,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SyncCheckpoint {
    pub generation: u64,
    pub cursor: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(tag = "command", rename_all = "snake_case")]
pub enum SyncAction {
    WriteLocal {
        action_id: String,
        depends_on: Vec<String>,
        reason: SyncPlanReason,
        target: SyncObjectRef,
        payload_revision: String,
        expected_local: ExpectedObjectState,
        expected_path_owner: ExpectedObjectState,
    },
    DeleteLocal {
        action_id: String,
        depends_on: Vec<String>,
        reason: SyncPlanReason,
        target: SyncObjectRef,
        expected_local: ExpectedObjectState,
        expected_path_owner: ExpectedObjectState,
    },
    MoveLocal {
        action_id: String,
        depends_on: Vec<String>,
        reason: SyncPlanReason,
        source: SyncObjectRef,
        target_path: String,
        expected_source_owner: ExpectedObjectState,
        expected_target_owner: ExpectedObjectState,
    },
    PutRemote {
        action_id: String,
        depends_on: Vec<String>,
        reason: SyncPlanReason,
        target: SyncObjectRef,
        payload_revision: String,
        expected_remote: ExpectedObjectState,
        expected_local: ExpectedObjectState,
        idempotency_key: String,
    },
    DeleteRemote {
        action_id: String,
        depends_on: Vec<String>,
        reason: SyncPlanReason,
        target: SyncObjectRef,
        expected_remote: ExpectedObjectState,
        expected_local: ExpectedObjectState,
        idempotency_key: String,
    },
    MoveRemote {
        action_id: String,
        depends_on: Vec<String>,
        reason: SyncPlanReason,
        source: SyncObjectRef,
        target_path: String,
        expected_source_owner: ExpectedObjectState,
        expected_target_owner: ExpectedObjectState,
        expected_local: ExpectedObjectState,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        revision_from_dependency: Option<String>,
        idempotency_key: String,
    },
    RecordConflict {
        action_id: String,
        depends_on: Vec<String>,
        reason: SyncPlanReason,
        identity: String,
        entity: SyncObjectKind,
        local: ExpectedObjectState,
        remote: ExpectedObjectState,
        conflict_kind: ConflictKind,
    },
    ClearConflict {
        action_id: String,
        depends_on: Vec<String>,
        reason: SyncPlanReason,
        identity: String,
        entity: SyncObjectKind,
        expected_local: ExpectedObjectState,
        expected_remote: ExpectedObjectState,
    },
    AdvanceCheckpoint {
        action_id: String,
        depends_on: Vec<String>,
        reason: SyncPlanReason,
        expected: SyncCheckpoint,
        next: SyncCheckpoint,
    },
}

impl SyncAction {
    pub fn action_id(&self) -> &str {
        match self {
            Self::WriteLocal { action_id, .. }
            | Self::DeleteLocal { action_id, .. }
            | Self::MoveLocal { action_id, .. }
            | Self::PutRemote { action_id, .. }
            | Self::DeleteRemote { action_id, .. }
            | Self::MoveRemote { action_id, .. }
            | Self::RecordConflict { action_id, .. }
            | Self::ClearConflict { action_id, .. }
            | Self::AdvanceCheckpoint { action_id, .. } => action_id,
        }
    }

    pub fn depends_on(&self) -> &[String] {
        match self {
            Self::WriteLocal { depends_on, .. }
            | Self::DeleteLocal { depends_on, .. }
            | Self::MoveLocal { depends_on, .. }
            | Self::PutRemote { depends_on, .. }
            | Self::DeleteRemote { depends_on, .. }
            | Self::MoveRemote { depends_on, .. }
            | Self::RecordConflict { depends_on, .. }
            | Self::ClearConflict { depends_on, .. }
            | Self::AdvanceCheckpoint { depends_on, .. } => depends_on,
        }
    }
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ConflictKind {
    BothChanged,
    DeleteVsChange,
    PathOccupied,
    Rejected,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct MirrorPlanIssue {
    pub code: String,
    pub message: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub path: Option<String>,
    pub blocking: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct MirrorPlanSummary {
    pub uploads: usize,
    pub downloads: usize,
    pub conflicts: usize,
    pub blocking_issues: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct MirrorSyncPlan {
    pub plan_version: u32,
    pub engine_profile: String,
    pub protocol_profile: String,
    pub planner_policy: String,
    pub projection_policy: String,
    pub fingerprint: String,
    pub replica_id: String,
    pub mode: SyncReplicaMode,
    pub kind: String,
    pub base_cursor: Option<u64>,
    pub authority_cursor: u64,
    pub scope_epoch: u64,
    pub checkpoint_generation: u64,
    pub selective_sync: SelectiveSyncPolicy,
    pub actions: Vec<SyncAction>,
    pub issues: Vec<MirrorPlanIssue>,
    pub summary: MirrorPlanSummary,
}
